nutils/expression_v2.py

```diff
@@ -792,20 +791,24 @@ def jump(self, array: function.Array) -> function.Array:
     return function.jump(array)
 
   def add(self, *args: Tuple[bool, function.Array]) -> function.Array:
-    assert all(arg.shape == args[0][1].shape for neg, arg in args[1:])
+    assert all(numpy.shape(arg) == numpy.shape(args[0][1]) for neg, arg in args[1:])
     negated = (-arg if neg else arg for neg, arg in args)
-    return functools.reduce(function.add, negated)
+    return functools.reduce(numpy.add, negated)
+
+  def append_axes(self, array, shape):
+    shuffle = numpy.concatenate([len(shape) + numpy.arange(numpy.ndim(array)), numpy.arange(len(shape))])
+    return numpy.transpose(numpy.broadcast_to(array, shape + numpy.shape(array)), shuffle)
 
   def multiply(self, *args: function.Array) -> function.Array:
     result = args[0]
     for arg in args[1:]:
-      result = function.multiply(function._append_axes(result, arg.shape), function._prepend_axes(arg, result.shape))
+      result = numpy.multiply(self.append_axes(result, numpy.shape(arg)), arg)
     return result
 
   def divide(self, numerator: function.Array, denominator: function.Array) -> function.Array:
-    assert denominator.ndim == 0
-    return function.divide(numerator, function._append_axes(denominator, numerator.shape))
+    assert numpy.ndim(denominator) == 0
+    return numpy.true_divide(numerator, numpy.broadcast_to(denominator, numpy.shape(denominator) + numpy.shape(numerator)))
 
   def power(self, base: function.Array, exponent: function.Array) -> function.Array:
-    assert exponent.ndim == 0
-    return function.power(base, function._append_axes(exponent, base.shape))
+    assert numpy.ndim(exponent) == 0
+    return numpy.power(base, numpy.broadcast_to(exponent, numpy.shape(exponent) + numpy.shape(base)))
```

Comment:
`exponent` has no shape, so you can simply use `numpy.broadcast_to(exponent, numpy.shape(base))`. Or let `numpy.power` do the magic.

Reply:
fixed